internal/cmdsearchreplace/cmdsearchreplace_test.go

```diff
@@ -0,0 +1,453 @@
+spec:
+  replicas: 3
+  foo: 3
+---
+apiVersion: apps/v1
+kind: Service
+metadata:
+  name: nginx-service
+ `,
+			expectedResources: `
+apiVersion: apps/v1
+kind: Deployment
+metadata:
+  name: nginx-deployment
+spec:
+  replicas: 4
+  foo: 3
+---
+apiVersion: apps/v1
+kind: Service
+metadata:
+  name: nginx-service
+ `,
+		},
+		{
+			name: "error when both by-value and by-regex provided",
+			args: []string{"--by-value", "nginx-deployment", "--by-value-regex", "nginx-*"},
+			input: `
+apiVersion: apps/v1
+kind: Deployment
+metadata:
+  name: nginx-deployment
+spec:
+  replicas: 3
+ `,
+			expectedResources: `
+apiVersion: apps/v1
+kind: Deployment
+metadata:
+  name: nginx-deployment
+spec:
+  replicas: 3
+ `,
+			errMsg: `only one of ["by-value", "by-value-regex"] can be provided`,
+		},
+	}
+	for i := range tests {
+		test := tests[i]
+		t.Run(test.name, func(t *testing.T) {
+			// reset the openAPI afterward
+
+			baseDir, err := ioutil.TempDir("", "")
+			if !assert.NoError(t, err) {
+				t.FailNow()
+			}
+			defer os.RemoveAll(baseDir)
+
+			r, err := ioutil.TempFile(baseDir, "k8s-cli-*.yaml")
+			if !assert.NoError(t, err) {
+				t.FailNow()
+			}
+			defer os.Remove(r.Name())
+			err = ioutil.WriteFile(r.Name(), []byte(test.input), 0600)
+			if !assert.NoError(t, err) {
+				t.FailNow()
+			}
+
+			runner := NewSearchRunner("")
+			out := &bytes.Buffer{}
+			runner.Command.SetOut(out)
+			runner.Command.SetArgs(append([]string{baseDir}, test.args...))
+			err = runner.Command.Execute()
+			if test.errMsg != "" {
+				if !assert.NotNil(t, err) {
+					t.FailNow()
+				}
+				if !assert.Contains(t, err.Error(), test.errMsg) {
+					t.FailNow()
+				}
+			}
+
+			if test.errMsg == "" && !assert.NoError(t, err) {
+				t.FailNow()
+			}
+
+			// normalize path format for windows
+			actualNormalized := strings.ReplaceAll(
+				strings.ReplaceAll(out.String(), "\\", "/"),
+				"//", "/")
+
+			expected := strings.ReplaceAll(test.out, "${baseDir}", baseDir)
+			expected = strings.ReplaceAll(expected, "${filePath}", filepath.Base(r.Name()))
+			expectedNormalized := strings.ReplaceAll(
+				strings.ReplaceAll(expected, "\\", "/"),
+				"//", "/")
+
+			if test.errMsg == "" && !assert.Equal(t, expectedNormalized, actualNormalized) {
+				t.FailNow()
+			}
+
+			actualResources, err := ioutil.ReadFile(r.Name())
+			if !assert.NoError(t, err) {
+				t.FailNow()
+			}
+			if !assert.Equal(t,
+				strings.TrimSpace(test.expectedResources),
+				strings.TrimSpace(string(actualResources))) {
+				t.FailNow()
+			}
+		})
+	}
+}
+
+func TestSearchSubPackages(t *testing.T) {
+	var tests = []struct {
+		name    string
+		dataset string
+		args    []string
+		out     string
+		errMsg  string
+	}{
+		{
+			name:    "search-replace-recurse-subpackages",
+			dataset: "dataset-with-autosetters",
+			args:    []string{"--by-value", "myspace", "--put-literal", "otherspace"},
+			out: `${baseDir}/
+matched 0 field(s)
+
+${baseDir}/mysql/
```

Comment:
Not really specific to this feature, but I would like to see a short design doc on how we should think about the UX when running a command on subpackages. We really need consistence across all commands so I think having a set of guidelines could be useful.

Reply:
+1, I don't know where we stand with recursive behavior across features.

Reply:
Agree. I followed the pattern followed by `cfg` commands released recently.